[clang] Fix PointerAuth semantics of cpp_trivially_relocatable

ojhunt opened the pull request:
This adds a function to ASTContext to query whether a type contains values with address discriminated pointer auth, and performs the required semantic checks to ensure correct reporting of relocatablity in those cases.

For the standardized version, __builtin_is_cpp_trivially_relocatable this means rejecting unions of types containing address discriminated values.

For the old deprecated __builtin_is_trivially_relocatable this means rejecting any type containing an address discriminated value.

This PR does not update the codegen for __builtin_trivially_relocate, that will be in a follow on PR that is much more complex.

ojhunt committed:
[clang] Fix PointerAuth semantics of cpp_trivially_relocatable

This adds a function to ASTContext to query whether a type contains
values with address discriminated pointer auth, and performs the
required semantic checks to ensure correct reporting of relocatablity
in those cases.

For the standardized version, __builtin_is_cpp_trivially_relocatable
this means rejecting unions of types containing address discriminated
values.

For the old deprecated __builtin_is_trivially_relocatable this means
rejecting any type containing an address discriminated value.

This PR does not update the codegen for __builtin_trivially_relocate,
that will be in a follow on PR that is much more complex.

cor3ntin commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -1705,6 +1705,40 @@ void ASTContext::setRelocationInfoForCXXRecord(
+    return false;
+
+  auto SaveReturn = [this, RD](bool Result) {
+    RecordContainsAddressDiscriminatedPointerAuth.insert({RD, Result});
```
Can you move `SaveReturn` after the map lookup?

ojhunt replied:
llvm's map does not have emplace afaict

cor3ntin commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -628,10 +628,13 @@ class ASTContext : public RefCountedBase<ASTContext> {
 private:
   llvm::DenseMap<const CXXRecordDecl *, CXXRecordDeclRelocationInfo>
       RelocatableClasses;
+  llvm::DenseMap<const RecordDecl *, bool>
+      RecordContainsAddressDiscriminatedPointerAuth;
```
We should do a better job at documenting these

cor3ntin replied:
Also, maybe add a fixme comment along the lines of "//FIXME store in RecordDeclBitfields?"

ojhunt replied:
Done and done.

cor3ntin reviewed (approved):
LGTM modulo nits

ojhunt committed:
Add comments and more tests

ojhunt commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -15121,6 +15155,21 @@ ASTContext::baseForVTableAuthentication(const CXXRecordDecl *ThisClass) {
   return PrimaryBase;
 }
 
+bool ASTContext::hasAddressDiscriminatedVTableAuthentication(
```
Which function?

ojhunt reviewed (commented):
goddammit review comment mode :D

ojhunt committed:
Address corentin comments

ojhunt committed:
__unused doesn't work on the linux and windows bots

ojhunt commented:
Very glad I didn't just assume my local build was sufficient and waited for the bots :D

ojhunt commented:
Actually I've given this a lot of thought, I think we should not consider explicitly __ptrauth qualified types to be relocatable, how does that sound?

ojhunt commented:
found how to unmark the approval as the next update will change this fairly substantially

ojhunt committed:
Address discriminated __ptrauth should not be relocatable

After consideration, while "supportable" by relocation, it
seems much more reasonable for people to understand that
explicitly qualified types can have different relocatability
semantics.

Implicitly address discriminated fields should be relocatable
as developers in general cannot control those behaviours, hence
the trivial_relocation language explicitly acknowledges that
this impacts unions.

Explicitly address discriminated fields do however drastically
change the developer aware model of a type, and the practical
requirements for efficiently handling __ptrauth qualified types
mean that you would want

   some_addr_discriminated_t[100];

to not be reported as trivially relocatable, while

   struct {
     ...
     some_addr_discriminated_t field;
     ...
   }

to be reported as such, and the actual decision would regress
to what proportion of the object being relocated is subject to
such a policy.

Having is_trivially_relocatable change behaviour according to
some hypothetical internal heuristic is clearly a terrible
idea, and so I think for now we should instead simply limit
the application of trivial relocation to address discriminated
data to those fields that outside of the control of developers.

ojhunt committed:
sigh, formatting

cor3ntin commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -629,25 +629,46 @@ class ASTContext : public RefCountedBase<ASTContext> {
+  };
+
+  // A simple helper function to short circuit pointer auth checks.
+  bool isPointerAuthenticationAvailable() const {
+    return LangOpts.PointerAuthCalls || LangOpts.PointerAuthIntrinsics ||
+           LangOpts.PointerAuthVTPtrAddressDiscrimination;
+  }
```
2 questions; Why do we have 3 options, and why do we care about anything but `PointerAuthVTPtrAddressDiscrimination`here?

ojhunt replied:
The first is historical sadness, we can probably make this better when I move the pointer auth options and similar into the target info rather than the obtuse combination of codegen and langopts we currently have. When we do that we can probably just have a single "the target supports pointer auth" flag which is what this is trying to approximate.

The reason that `PointerAuthVTPtrAddressDiscrimination` is not sufficient is that there is a qualifier that can be specified explicitly.

That said as we discussed (for people following at home @cor3ntin and I are talking at wg21) if someone had `PointerAuthVTPtrAddressDiscrimination` enabled without `PointerAuthCalls` and `PointerAuthIntrinsics` would likely be in a world of hurt, so I'll remove that separate check.

ojhunt committed:
remove extraneous check
